examples/chat/src/network/discovery.rs

```diff
@@ -0,0 +1,405 @@
+            }
+        }
+
+        self.discovery_state.clear_pending_addr_changes();
+
+        Ok(())
+    }
+
+    // Updates rendezvous registration on the remote rendezvous peer.
+    // If there are no external addresses for the node, the registration is considered successful.
+    // This function expectes that the relay peer is already connected.
+    pub(crate) fn update_rendezvous_registration(
+        &mut self,
+        rendezvous_peer: &PeerId,
+    ) -> eyre::Result<()> {
+        if let Err(err) = self.swarm.behaviour_mut().rendezvous.register(
+            self.rendezvous_namespace.clone(),
+            *rendezvous_peer,
+            None,
+        ) {
+            match err {
+                libp2p::rendezvous::client::RegisterError::NoExternalAddresses => {}
+                err => eyre::bail!(err),
+            }
+        }
+
+        debug!(
+            %rendezvous_peer, rendezvous_namespace=%(self.rendezvous_namespace),
+            "Sent register request to rendezvous node"
+        );
+        Ok(())
+    }
+
+    // Creates relay reservation if node doesn't have a relayed address on the relay peer.
+    // This function expectes that the relay peer is already connected.
+    pub(crate) fn create_relay_reservation(&mut self, relay_peer: &PeerId) -> eyre::Result<()> {
+        let peer_info = self
+            .discovery_state
+            .get_peer_info(relay_peer)
+            .wrap_err("Failed to get peer info")?;
+
+        let external_addrs = self
+            .swarm
+            .external_addresses()
+            .filter(|addr| addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)))
+            .collect::<HashSet<_>>();
+
+        let preferred_addr = peer_info
+            .get_preferred_addr()
+            .wrap_err("Failed to get preferred addr for relay peer")?;
+
+        let relayed_addr = match preferred_addr
+            .clone()
+            .with(multiaddr::Protocol::P2pCircuit)
+            .with_p2p(self.swarm.local_peer_id().clone())
+        {
+            Ok(addr) => addr,
+            Err(err) => {
+                eyre::bail!("Failed to construct relayed addr for relay peer: {:?}", err)
+            }
+        };
+        let is_relay_reservation_required = !(matches!(
+            peer_info.relay_reservation_status(),
+            Some(RelayReservationStatus::Requested)
+        ) || external_addrs.contains(&relayed_addr));
+
+        debug!(
+            ?peer_info,
+            ?external_addrs,
+            %is_relay_reservation_required,
+            "Checking if relay reservation is required"
+        );
+
+        if !is_relay_reservation_required {
+            return Ok(());
+        }
+
+        self.swarm.listen_on(relayed_addr)?;
+        self.discovery_state
+            .update_relay_reservation_status(&relay_peer, RelayReservationStatus::Requested)?;
+
+        Ok(())
+    }
+}
+
+// The rendezvous protocol name is not public in libp2p, so we have to define it here.
```

Comment:
This looks dangerous

Reply:
I don't think as well they will change it but if they do, do we have easy way to detect it? If you misspell the name, what is the error?